File: .claude/skills/spec-check/scripts/spec_check/invariants/fr_coverage.py
"""P2 — every `…-fr-…` id defined in the PRD is claimed by exactly one slice's
`**Traces to**:` block, and no slice traces to an id the PRD never defines.

Port of `tools/spec-check/src/invariants/fr_coverage.rs`.
"""

import re

from ..corpus import split_lines
from ..finding import Finding, Severity
from .closure import is_design_slice

_DEFINED_ID = re.compile(r"\*\*ID\*\*:\s*`(cpt-cf-[a-z0-9-]*-fr-[a-z0-9-]+)`")
_ID = re.compile(r"`(cpt-cf-[a-z0-9-]*-fr-[a-z0-9-]+)`")

_DIRECTLY_ADDRESSES = "This slice directly addresses:"


def _defined_requirements(corpus):
    text = corpus.text("PRD.md")
    if text is None:
        return []
    return sorted({m.group(1) for m in _DEFINED_ID.finditer(text)})


def _collect_traces_to(path, text, out):
    """Ids inside a `**Traces to**:` block, which runs until the first blank line.

    A set per id, not a list: a copy/paste or line-wrap slip can repeat the same
    id in one block, and that is one dangling defect, not one per repetition.
    Returns whether the marker was seen at all — the caller needs both collectors'
    "did I see my marker" signal to tell a gear using neither convention from one
    using this convention exhaustively.
    """
    in_block = False
    seen = False
    for line in split_lines(text):
        if "**Traces to**:" in line:
            in_block = True
            seen = True
        elif in_block and line.strip() == "":
            in_block = False
        if in_block:
            for match in _ID.finditer(line):
                out.setdefault(match.group(1), set()).add(path)
    return seen


def _collect_directly_addresses(path, text, out):
    """Ids inside a `This slice directly addresses:` block.

    Unlike `**Traces to**:`, the marker is followed by a blank line before the
    bullets start (design/01-foundation.md's only occurrence: marker, blank line,
    six bullets, then end of file, no trailing heading) — so "stop at the first
    blank line" would collect nothing here. Instead: skip blank lines after the
    marker, collect ids from bullet lines, and stop at the next heading or the
    first non-blank, non-bullet line. Falling off the end closes the block too.
    """
    in_block = False
    seen = False
    for line in split_lines(text):
        if _DIRECTLY_ADDRESSES in line:
            in_block = True
            seen = True
            continue
        if not in_block:
            continue
        trimmed = line.lstrip()
        if trimmed == "":
            continue  # blank line between the marker and the bullets: stay open
        if line.startswith("#") or not trimmed.startswith("-"):
            in_block = False
            continue
        for match in _ID.finditer(line):
            out.setdefault(match.group(1), set()).add(path)
    return seen


def _claimed_requirements(corpus):
    """Collects ids claimed by slice documents, recognising the two conventions
    the design set actually uses, and returns `(claims, findings, convention_seen)`.

    Eleven of the twelve pricing slices use `**Traces to**:`; the twelfth
    (design/01-foundation.md) predates that convention and lists requirements
    under `This slice directly addresses:`. Both are equally valid ownership
    claims — the P2 question is "is this requirement claimed by a slice", not
    "claimed in a particular markdown shape" — so both are read. A document using
    the second shape still gets a divergence finding of its own: the checker must
    not silently absorb a shape split from the rest of the set.

    Scoped to `is_design_slice` documents only. Without this, a stray
    `**Traces to**:` in a non-slice document would count towards
    `convention_seen` (wrongly reviving the per-id sweep for a gear whose real
    slices use neither known convention) and could even wrongly "claim" a
    requirement no real slice claims.
    """
    out = {}
    findings = []
    convention_seen = False

    for path, text in corpus.files():
        if not is_design_slice(path):
            continue
        saw_traces_to = _collect_traces_to(path, text, out)
        saw_directly_addresses = _collect_directly_addresses(path, text, out)
        if saw_directly_addresses:
            findings.append(Finding(
                "P2/traceability-convention-divergent",
                Severity.LOW,
                path,
                None,
                "{} claims requirements via `This slice directly addresses:`; "
                "the rest of the design set uses `**Traces to**:`".format(path),
            ))
        convention_seen = convention_seen or saw_traces_to or saw_directly_addresses

    return out, findings, convention_seen


def check(corpus):
    """P2 over one corpus.

    Exactly one claiming slice, not at least one: the map has always been
    id -> *set of files*, but nothing ever reported `len() > 1`, so the ownership
    ambiguity this invariant exists to catch — two slices both asserting they
    implement one requirement, with no single owner to change when it moves — was
    live and undetected. Both halves are Low: they are the two directions of the
    same cardinality question, and neither proves a document states something
    false the way `fr-dangling` (Medium) does.

    A gear whose slices use *neither* recognised convention is not "fully
    uncovered" — it is unparsed. Emitting a `P2/fr-unclaimed` for every one of
    that gear's requirements would report confident-sounding gaps P2 has no basis
    for, which is worse than silence. So it emits exactly one
    `P2/traceability-convention-unknown` instead of the per-id sweep.
    `P2/fr-dangling` is unaffected either way.
    """
    defined = _defined_requirements(corpus)
    claimed, findings, convention_seen = _claimed_requirements(corpus)

    if convention_seen:
        for ident in defined:
            if ident not in claimed:
                findings.append(Finding(
                    "P2/fr-unclaimed",
                    Severity.LOW,
                    "PRD.md",
                    None,
                    "{} is claimed by no slice's Traces-to line".format(ident),
                ))
    else:
        findings.append(Finding(
            "P2/traceability-convention-unknown",
            Severity.LOW,
            "PRD.md",
            None,
            "P2 cannot verify requirement coverage for {}: no slice uses a recognised "
            "traceability convention (`**Traces to**:` or `This slice directly "
            "addresses:`) — {} requirement(s) went unchecked as a result; per-id claims "
            "are not reported for this gear rather than reporting every requirement as "
            "unclaimed".format(corpus.root(), len(defined)),
        ))

    defined_set = set(defined)
    for ident in sorted(claimed):
        files = sorted(claimed[ident])
        # Reported whether or not the PRD defines the id: two slices both claiming
        # an id is an ownership defect on its own, and an id that is *also*
        # dangling gets both findings because they are two different defects in
        # two different documents.
        if len(files) > 1:
            findings.append(Finding(
                "P2/fr-multiply-claimed",
                Severity.LOW,
                "PRD.md",
                None,
                "{} is claimed by {} slices, not one: {}".format(
                    ident, len(files), ", ".join(files)
                ),
            ))
        if ident not in defined_set:
            for path in files:
                findings.append(Finding(
                    "P2/fr-dangling",
                    Severity.MEDIUM,
                    path,
                    None,
                    "Traces to {}, which the PRD does not define".format(ident),
                ))

    return findings

File: tools/spec-check/src/invariants/fr_coverage.rs
use std::collections::{BTreeMap, BTreeSet};

use regex::Regex;

use crate::Corpus;
use crate::finding::{Finding, Severity};

/// P2 — every `…-fr-…` id defined in the PRD is claimed by at least one slice's
/// `**Traces to**:` block, and no slice traces to an id the PRD never defines.
pub fn check(corpus: &Corpus) -> Vec<Finding> {
    let defined = defined_requirements(corpus);
    let (claimed, mut findings) = claimed_requirements(corpus);

    for id in &defined {
        if !claimed.contains_key(id) {
            findings.push(Finding {
                invariant: "P2/fr-unclaimed".to_string(),
                severity: Severity::Low,
                file: "PRD.md".to_string(),
                line: None,
                message: format!("{id} is claimed by no slice's Traces-to line"),
            });
        }
    }

    for (id, files) in &claimed {
        if !defined.contains(id) {
            for f in files {
                findings.push(Finding {
                    invariant: "P2/fr-dangling".to_string(),
                    severity: Severity::Medium,
                    file: f.clone(),
                    line: None,
                    message: format!("Traces to {id}, which the PRD does not define"),
                });
            }
        }
    }

    findings
}

fn defined_requirements(corpus: &Corpus) -> BTreeSet<String> {
    let id =
        Regex::new(r"\*\*ID\*\*:\s*`(cpt-cf-[a-z0-9-]*-fr-[a-z0-9-]+)`").expect("valid id regex");
    corpus
        .text("PRD.md")
        .map(|t| id.captures_iter(t).map(|c| c[1].to_string()).collect())
        .unwrap_or_default()
}

/// Collects ids claimed by slice documents, recognising the two conventions the
/// design set actually uses. Eleven of the twelve pricing slices use
/// `**Traces to**:`; the twelfth (`design/01-foundation.md`) predates that
/// convention and instead lists requirements under a `This slice directly
/// addresses:` heading. Both are equally valid ownership claims — the P2
/// question is "is this requirement claimed by a slice", not "claimed in a
/// particular markdown shape" — so both are read for claims. A document using
/// the second shape still gets a `P2/traceability-convention-divergent`
/// finding of its own: the checker must not silently absorb a shape split from
/// the rest of the set.
fn claimed_requirements(corpus: &Corpus) -> (BTreeMap<String, BTreeSet<String>>, Vec<Finding>) {
    let id = Regex::new(r"`(cpt-cf-[a-z0-9-]*-fr-[a-z0-9-]+)`").expect("valid id regex");
    let mut out: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    let mut findings = Vec::new();

    for (path, text) in corpus.files() {
        if path == "PRD.md" {
            continue;
        }
        collect_traces_to(path, text, &id, &mut out);
        if collect_directly_addresses(path, text, &id, &mut out) {
            findings.push(Finding {
                invariant: "P2/traceability-convention-divergent".to_string(),
                severity: Severity::Low,
                file: path.to_string(),
                line: None,
                message: format!(
                    "{path} claims requirements via `This slice directly addresses:`; \
                     the rest of the design set uses `**Traces to**:`"
                ),
            });
        }
    }
    (out, findings)
}

/// Ids inside a `**Traces to**:` block, which runs until the first blank line.
/// A `BTreeSet` per id, not a `Vec`: a copy/paste or line-wrap slip can repeat
/// the same id in one block, and that is one dangling defect, not one per
/// repetition — deduplicating here (rather than in the dangling loop) keeps
/// both collectors and their caller honest about "which documents claim this
/// id", with no separate dedup step to forget.
fn collect_traces_to(
    path: &str,
    text: &str,
    id: &Regex,
    out: &mut BTreeMap<String, BTreeSet<String>>,
) {
    let mut in_block = false;
    for line in text.lines() {
        if line.contains("**Traces to**:") {
            in_block = true;
        } else if in_block && line.trim().is_empty() {
            in_block = false;
        }
        if in_block {
            for c in id.captures_iter(line) {
                out.entry(c[1].to_string())
                    .or_default()
                    .insert(path.to_string());
            }
        }
    }
}

/// Ids inside a `This slice directly addresses:` block. Unlike `**Traces
/// to**:`, the marker is followed by a blank line before the bullets start
/// (`design/01-foundation.md`'s only occurrence: marker, blank line, six
/// bullets, then end of file, no trailing heading) — so "stop at the first
/// blank line" would collect nothing here. Instead: skip blank lines after the
/// marker, collect ids from bullet lines (optional leading whitespace then
/// `-`), and stop at the next heading (`^#`) or the first non-blank,
/// non-bullet line — falling off the end of the document closes the block
/// too. Returns whether the marker was seen at all, so the caller can raise
/// the convention-divergence finding.
fn collect_directly_addresses(
    path: &str,
    text: &str,
    id: &Regex,
    out: &mut BTreeMap<String, BTreeSet<String>>,
) -> bool {
    const MARKER: &str = "This slice directly addresses:";
    let mut in_block = false;
    let mut seen = false;
    for line in text.lines() {
        if line.contains(MARKER) {
            in_block = true;
            seen = true;
            continue;
        }
        if !in_block {
            continue;
        }
        let trimmed = line.trim_start();
        if trimmed.is_empty() {
            continue; // blank line between the marker and the bullets: stay open
        }
        if line.starts_with('#') || !trimmed.starts_with('-') {
            in_block = false;
            continue;
        }
        for c in id.captures_iter(line) {
            out.entry(c[1].to_string())
                .or_default()
                .insert(path.to_string());
        }
    }
    seen
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Corpus;
    use std::path::PathBuf;

    fn pricing() -> Corpus {
        let root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../gears/bss/pricing/docs");
        Corpus::load(&root).expect("pricing corpus loads")
    }

    #[test]
    fn every_pricing_fr_is_claimed_by_a_slice() {
        let findings = check(&pricing());
        let orphans: Vec<_> = findings
            .iter()
            .filter(|f| f.invariant == "P2/fr-unclaimed")
            .collect();
        assert!(orphans.is_empty(), "unexpected: {orphans:#?}");
    }

    #[test]
    fn flags_an_fr_no_slice_traces_to() {
        let corpus = Corpus::from_parts(
            "synthetic",
            [
                (
                    "PRD.md",
                    "- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-lonely`\n- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-other`\n",
                ),
                ("design/01-a.md", "**Traces to**: `cpt-cf-bss-x-fr-other`\n"),
            ],
        );
        let findings = check(&corpus);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].invariant, "P2/fr-unclaimed");
        assert!(findings[0].message.contains("cpt-cf-bss-x-fr-lonely"));
    }

    #[test]
    fn reads_traces_to_lines_that_wrap() {
        let corpus = Corpus::from_parts(
            "synthetic",
            [
                (
                    "PRD.md",
                    "- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-first`\n- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-wrapped`\n",
                ),
                (
                    "design/01-a.md",
                    "**Traces to**: `cpt-cf-bss-x-fr-first`,\n`cpt-cf-bss-x-fr-wrapped`\n\nNext paragraph.\n",
                ),
            ],
        );
        assert!(check(&corpus).is_empty());
    }

    #[test]
    fn recognises_this_slice_directly_addresses_as_an_alternate_claim_convention() {
        // Mirrors design/01-foundation.md's exact shape: marker line, then a
        // *blank* line, then bullets, then end of file with no trailing
        // heading. A fixture whose bullets started immediately after the
        // marker would pass even with a wrong "stop at the first blank line"
        // rule (the `**Traces to**:` rule) — the blank line here is what
        // proves the stop condition is right, since that rule would collect
        // nothing at all from this shape.
        //
        // Two ids, not one: this is what makes "exactly one divergence
        // finding" a real assertion about the finding's cardinality (raised
        // once per document) rather than one that would also pass a latent
        // bug where the push lived inside the per-id capture loop instead of
        // after it (which would emit one divergence finding per id here).
        let corpus = Corpus::from_parts(
            "synthetic",
            [
                (
                    "PRD.md",
                    "- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-foundational`\n- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-alsofoundational`\n",
                ),
                (
                    "design/01-a.md",
                    "This slice directly addresses:\n\n- `cpt-cf-bss-x-fr-foundational` / `cpt-cf-bss-x-fr-alsofoundational` — does the thing\n",
                ),
            ],
        );
        let findings = check(&corpus);
        assert!(
            !findings.iter().any(|f| f.invariant == "P2/fr-unclaimed"),
            "requirements should be treated as claimed: {findings:#?}"
        );
        let divergences: Vec<_> = findings
            .iter()
            .filter(|f| f.invariant == "P2/traceability-convention-divergent")
            .collect();
        assert_eq!(divergences.len(), 1, "unexpected: {findings:#?}");
        assert_eq!(divergences[0].severity, Severity::Low);
        assert_eq!(divergences[0].file, "design/01-a.md");
    }

    #[test]
    fn flags_a_slice_tracing_to_a_requirement_that_does_not_exist() {
        let corpus = Corpus::from_parts(
            "synthetic",
            [
                ("PRD.md", "- [ ] `p1` - **ID**: `cpt-cf-bss-x-fr-real`\n"),
                (
                    "design/01-a.md",
                    "**Traces to**: `cpt-cf-bss-x-fr-real`, `cpt-cf-bss-x-fr-ghost`\n",
                ),
            ],
        );
        let findings = check(&corpus);
        assert_eq!(findings.len(), 1);
        assert_eq!(findings[0].invariant, "P2/fr-dangling");
        assert_eq!(findings[0].file, "design/01-a.md");
    }

    #[test]
    fn deduplicates_a_dangling_id_repeated_in_one_document() {
        // A copy/paste or line-wrap slip can list the same undefined id twice in
        // one Traces-to block. That is one defect, not two — repeated mentions of
        // the same (file, id) pair must not inflate the finding count.
        let corpus = Corpus::from_parts(
            "synthetic",
            [
                ("PRD.md", ""),
                (
                    "design/01-a.md",
                    "**Traces to**: `cpt-cf-bss-x-fr-ghost`, `cpt-cf-bss-x-fr-ghost`\n",
                ),
            ],
        );
        let findings = check(&corpus);
        assert_eq!(findings.len(), 1, "unexpected: {findings:#?}");
        assert_eq!(findings[0].invariant, "P2/fr-dangling");
        assert_eq!(findings[0].file, "design/01-a.md");
    }
}
